librad/src/identities/git.rs

```diff
@@ -0,0 +1,763 @@
+        let author = self.repo.signature()?;
+
+        self.repo
+            .commit(
+                None,
+                &author,
+                &author,
+                &sign::CommitMessage::new(message, signatures).to_string(),
+                &tree,
+                parents.iter().collect::<Vec<_>>().as_slice(),
+            )
+            .map(ContentId::from)
+    }
+}
+
+impl<'a> Git<'a, User> {
+    /// Attempt to read a [`User`] from commit `oid`, without verification.
+    pub fn get(&self, oid: git2::Oid) -> Result<User, error::Load> {
+        self.get_generic(oid)
+    }
+
+    /// Verify the user history with head commit `head`.
+    ///
+    /// The returned [`VerifiedUser`] is the **most recent** identity for which
+    /// the verification succeeded -- which may or may not be `head`.
+    pub fn verify(&self, head: git2::Oid) -> Result<VerifiedUser, error::VerifyUser> {
+        Ok(self.verify_generic(head)?)
+    }
+
+    /// Create a new [`User`] from a payload and delegations.
+    ///
+    /// The returned [`User`] (and the underlying commit) will not have any
+    /// parents, and will by signed by `signer`.
+    pub fn create<S>(
+        &self,
+        payload: UserPayload,
+        delegations: delegation::Direct,
+        signer: &S,
+    ) -> Result<User, error::Store<S::Error>>
+    where
+        S: Signer,
+        S::Error: std::error::Error + Send + Sync + 'static,
+    {
+        let doc = Doc {
+            version: 0,
+            replaces: None,
+            payload,
+            delegations: payload::UserDelegations::from(delegations),
+        };
+
+        let root: Revision = self.repo.blob(&Cjson(&doc).canonical_form()?)?.into();
+        let revision = {
+            let mut builder = self.repo.treebuilder(None)?;
+            builder.insert(root.to_string(), *root, 0o100_644)?;
+            builder.write().map(Revision::from)
+        }?;
+        let signatures = sign(signer, revision).map_err(error::Store::Signer)?.into();
+        let content_id = self.commit(
+            &format!("Initialised user identity {}", root),
+            &signatures,
+            revision,
+            &[],
+        )?;
+
+        Ok(Identity {
+            content_id,
+            root,
+            revision,
+            doc: doc.second(delegation::Direct::from),
+            signatures,
+        })
+    }
+
+    /// Update an existing [`SignedUser`] with a new payload and delegations.
+    ///
+    /// If both `payload` and `delegations` evaluate to `None`, or their values
+    /// result in the same revision as `base`, no new commit is made, and
+    /// the result is the unwrapped [`User`] of the `base` argument.
+    ///
+    /// Otherwise, the result is a new [`User`] whose parent is `base`.
+    pub fn update<S>(
+        &self,
+        base: SignedUser,
+        payload: impl Into<Option<UserPayload>>,
+        delegations: impl Into<Option<delegation::Direct>>,
```

Comment:
I find the option of not passing either of these a bit unergonomic. Unless there is some use case for passing neither and just getting back the `base`? :shrug: 

Would it make more sense to use something like `These` so that we have to update at least one if not both?

Reply:
Hmya, but can we make an ergonomic `These` in rrrrrst? Builder pattern maybe? :stuck_out_tongue:

Reply:
Shameful plug: https://crates.io/crates/these, I gave it a go when I was learning Rust for my interview here :sweat_smile:

Reply:
It's probably too rough around the edges anyway and this is certainly not a blocker

Reply:
It's also that we can't statically prevent that this doesn't end up with the same hash -- you could be passing non-`None` values, but they are the same as in `base`. I figured that this shouldn't be an error, but the method be idempotent -- you just get the same thing back, and proceed with your updates.

Reply:
Yup, that check made sense to me :+1: I was more just thinking of the ergonomics of at least attempting to update. No biggy anyway :)